### PR TITLE
[WIP] Swap to a random number generator that doesn't use atomics/synchronization on hot path

### PR DESCRIPTION
**Pull request by lukecwik**
The Java Random and ThreadLocalRandom are thread safe and use atomics and/or synchonization internally.

This is for #21250
 
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/data/PCollectionConsumerRegistry.java [hunk 1]
```diff
@@ -393,89 +394,58 @@ public double getProgress() {
     }
   }
 
-  private static class SampleByteSizeDistribution<T> {
+  public static class SampleByteSizeDistribution<T> extends ElementByteSizeObserver {
```
Comment (lukecwik) [inline sdks/java/harness/src/main/java/org/apache/beam/fn/harness/data/PCollectionConsumerRegistry.java:397]:
```suggestion
  private static class SampleByteSizeDistribution<T> extends ElementByteSizeObserver {
```

**Comment by lukecwik** [comment]
Run Java PreCommit

**Comment by lukecwik** [comment]
Run Java_PVR_Flink_Batch PreCommit

**Comment by lukecwik** [comment]
Run Python_Runners PreCommit

**Comment by lukecwik** [comment]
Run SQL_Java11 PreCommit

**Comment by lukecwik** [comment]
Run SQL_Java17 PreCommit

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/data/PCollectionConsumerRegistry.java [hunk 2]
```diff
@@ -393,89 +394,58 @@ public double getProgress() {
...
-              / Math.log(1.0 - RESERVOIR_SIZE / (double) samplingToken);
-      return samplingToken + (int) gap;
+      // We unconditionally sample the first samplingCutoff elements. For the
+      // next samplingCutoff elements, the sampling probability drops from 100%
+      // to 50%. The probability of sampling the Nth element is:
+      // min(1, samplingCutoff / N), with an additional lower bound of
+      // samplingCutoff / samplingTokenUpperBound. This algorithm may be refined
+      // later.
+      samplingToken = Math.min(samplingToken + 1, SAMPLING_TOKEN_UPPER_BOUND);
+      return randomGenerator.nextInt(samplingToken) < SAMPLING_CUTOFF;
```
Comment (robertwb) [inline sdks/java/harness/src/main/java/org/apache/beam/fn/harness/data/PCollectionConsumerRegistry.java:448]: Even if we switch to a thread-local RNG, I'm concerned that calling nextInt on every single value will be more expensive than doing the fast reservoir sampling algorithm that we had here before.

Comment (lukecwik) [reply sdks/java/harness/src/main/java/org/apache/beam/fn/harness/data/PCollectionConsumerRegistry.java:448]: Yeah this is still WIP and marked it as so. I'm still trying to root cause why the existing implementation shows up on FlexWordCount profiles.

**Pull request closed without merging**